### ISSUE TITLE
Cannot install or update any packages with 0.186.0 under Windows using proxy

### ISSUE THREAD

**Issue by weblogixx** [atom/atom#5913]
Hi all,

I have a problem installing and updating packages in the current Atom for Windows release.
There seems to be a problem running apm that seems to have been around for some time now (as it did work with the same settings before 0.180). My local node-Setup is just working fine with the same settings and toolchain, so that should not be the problem. It does not seem to be a general proxy problem as I can see available themes and packages, as well as updates in Atom.

My current setup / Toolchain (with proxy and username obusfacated):
```
Windows 7 (64 bit)

C:\Windows\system32>apm -v
apm  0.142.0
npm  2.5.1
node 0.10.35
python 2.7.9
git 1.9.4.msysgit.2
visual studio 2013

C:\Windows\system32>apm config list
; cli configs
globalconfig = "C:\\Users\\myuser\\.atom\\.apm\\.apmrc"
user-agent = "npm/2.5.1 node/v0.10.35 win32 ia32"
userconfig = "C:\\Users\\myuser\\.atom\\.apmrc"

; userconfig C:\Users\myuser\.atom\.apmrc
http-proxy = "http://proxy.company.com:8080"
https-proxy = "http://proxy.company.com:8080/"

; globalconfig C:\Users\christian.schilling\.atom\.apm\.apmrc
cache = "C:\\Users\\myuser\\.atom\\.apm"

; node bin location = c:\Users\myuser\AppData\Local\atom\app-0.186.0\resources\app\apm\bin\node.exe
; cwd = C:\Windows\system32
; HOME = C:\Users\myuser
; 'npm config ls -l' to show all defaults.
```

When I try to update or install packages I get the following error (even running a privileged cmd):
```
C:\Windows\system32>apm update
Package Updates Available (2)
├── docblockr 0.5.8 -> 0.6.3
└── minimap 4.1.3 -> 4.5.0

Would you like to install these updates? (yes)

Installing docblockr@0.6.3 to C:\Users\USERNAME\.atom\packages
gyp info it worked if it ends with ok
gyp info using node-gyp@1.0.2
gyp info using node@0.10.35 | win32 | ia32
gyp http GET https://atom.io/download/atom-shell/v0.21.0/node-v0.21.0.tar.gz
gyp WARN install got an error, rolling back install
gyp
```

When using the gui, I get the "Compiler tools not found" error:
```
C:\Windows\system32>apm install --check
Checking for native build tools failed

gyp info it worked if it ends with ok
gyp info using node-gyp@1.0.2
gyp info using node@0.10.35 | win32 | ia32
gyp http GET https://atom.io/download/atom-shell/v0.21.0/node-v0.21.0.tar.gz
gyp WARN install got an error, rolling back install
gyp
```

After this try I ran:
```
apm install --verbose
gyp info it worked if it ends with ok
gyp info using node-gyp@1.0.2
gyp info using node@0.10.35 | win32 | ia32
gyp http GET https://atom.io/download/atom-shell/v0.21.0/node-v0.21.0.tar.gz
REQUEST { proxy: 'http://proxy.company.com:8080/',
  headers: { 'User-Agent': 'node-gyp v1.0.2 (node v0.10.35)' },
  uri: 'https://atom.io/download/atom-shell/v0.21.0/node-v0.21.0.tar.gz',
  callback: undefined }
REQUEST make request https://atom.io/download/atom-shell/v0.21.0/node-v0.21.0.tar.gz
REQUEST onRequestResponse https://atom.io/download/atom-shell/v0.21.0/node-v0.21.0.tar.gz 302 { server: 'Cowboy',
  date: 'Tue, 10 Mar 2015 06:25:07 GMT',
  connection: 'keep-alive',
  'strict-transport-security': 'max-age=631152000',
  'x-frame-options': 'DENY',
  'x-xss-protection': '1; mode=block',
  'x-content-type-options': 'nosniff',
  'content-security-policy': 'default-src \'self\'; connect-src \'self\'; font-src \'self\'; frame-src \'self\'; img-src https://* \'self\' data:; media-src \'none\'; object-src \'self\'; script-src \
'self\' \'unsafe-inline\' https://ssl.google-analytics.com https://www.google-analytics.com; style-src \'self\' \'unsafe-inline\';',
  location: 'http://gh-contractor-zcbenz.s3.amazonaws.com/atom-shell/dist/v0.21.0/node-v0.21.0.tar.gz',
  'content-type': 'text/html; charset=utf-8',
  'cache-control': 'no-cache',
  'set-cookie': [ 'request_method=GET; path=/; secure' ],
  'x-request-id': 'e20b681a-e86f-43f9-87c5-b15c6359e2cd',
  'x-runtime': '0.017689',
  'transfer-encoding': 'chunked',
  via: '1.1 vegur' }
REQUEST redirect http://gh-contractor-zcbenz.s3.amazonaws.com/atom-shell/dist/v0.21.0/node-v0.21.0.tar.gz
REQUEST redirect to http://gh-contractor-zcbenz.s3.amazonaws.com/atom-shell/dist/v0.21.0/node-v0.21.0.tar.gz
REQUEST {}
REQUEST response end http://gh-contractor-zcbenz.s3.amazonaws.com/atom-shell/dist/v0.21.0/node-v0.21.0.tar.gz 302 { server: 'Cowboy',
  date: 'Tue, 10 Mar 2015 06:25:07 GMT',
  connection: 'keep-alive',
  'strict-transport-security': 'max-age=631152000',
  'x-frame-options': 'DENY',
  'x-xss-protection': '1; mode=block',
  'x-content-type-options': 'nosniff',
  'content-security-policy': 'default-src \'self\'; connect-src \'self\'; font-src \'self\'; frame-src \'self\'; img-src https://* \'self\' data:; media-src \'none\'; object-src \'self\'; script-src \
'self\' \'unsafe-inline\' https://ssl.google-analytics.com https://www.google-analytics.com; style-src \'self\' \'unsafe-inline\';',
  location: 'http://gh-contractor-zcbenz.s3.amazonaws.com/atom-shell/dist/v0.21.0/node-v0.21.0.tar.gz',
  'content-type': 'text/html; charset=utf-8',
  'cache-control': 'no-cache',
  'set-cookie': [ 'request_method=GET; path=/; secure' ],
  'x-request-id': 'e20b681a-e86f-43f9-87c5-b15c6359e2cd',
  'x-runtime': '0.017689',
  'transfer-encoding': 'chunked',
  via: '1.1 vegur' }
REQUEST make request http://gh-contractor-zcbenz.s3.amazonaws.com/atom-shell/dist/v0.21.0/node-v0.21.0.tar.gz
gyp WARN install got an error, rolling back install
gyp
```

The same last request line can also be found when trying to update. Interestingly the packages I want to update seem to be downloadable:
```
REQUEST onRequestResponse https://codeload.github.com/NikhilKalige/docblockr/legacy.tar.gz/v0.6.3 200 { 'content-length': '531353',
  'access-control-allow-origin': 'https://render.githubusercontent.com',
  'content-security-policy': 'default-src \'none\'',
  'x-xss-protection': '1; mode=block',
  'x-frame-options': 'deny',
  'x-content-type-options': 'nosniff',
  'strict-transport-security': 'max-age=31536000',
  vary: 'Authorization,Accept-Encoding',
  etag: '"12459b3f72507c12b6258f9bd550553fc6f43476"',
  'content-type': 'application/x-gzip',
  'content-disposition': 'attachment; filename=NikhilKalige-docblockr-v0.6.3-0-g12459b3.tar.gz',
  date: 'Tue, 10 Mar 2015 06:26:26 GMT' }
REQUEST finish init function https://codeload.github.com/NikhilKalige/docblockr/legacy.tar.gz/v0.6.3
REQUEST response end https://codeload.github.com/NikhilKalige/docblockr/legacy.tar.gz/v0.6.3 200 { 'content-length': '531353',
  'access-control-allow-origin': 'https://render.githubusercontent.com',
  'content-security-policy': 'default-src \'none\'',
  'x-xss-protection': '1; mode=block',
  'x-frame-options': 'deny',
  'x-content-type-options': 'nosniff',
  'strict-transport-security': 'max-age=31536000',
  vary: 'Authorization,Accept-Encoding',
  etag: '"12459b3f72507c12b6258f9bd550553fc6f43476"',
  'content-type': 'application/x-gzip',
  'content-disposition': 'attachment; filename=NikhilKalige-docblockr-v0.6.3-0-g12459b3.tar.gz',
  date: 'Tue, 10 Mar 2015 06:26:26 GMT' }
gyp info it worked if it ends with ok
gyp info using node-gyp@1.0.2
gyp info using node@0.10.35 | win32 | ia32
gyp http GET https://atom.io/download/atom-shell/v0.21.0/node-v0.21.0.tar.gz
REQUEST { proxy: 'http://proxy.company.com:8080/',
  headers: { 'User-Agent': 'node-gyp v1.0.2 (node v0.10.35)' },
  uri: 'https://atom.io/download/atom-shell/v0.21.0/node-v0.21.0.tar.gz',
  callback: undefined }
REQUEST make request https://atom.io/download/atom-shell/v0.21.0/node-v0.21.0.tar.gz
REQUEST onRequestResponse https://atom.io/download/atom-shell/v0.21.0/node-v0.21.0.tar.gz 302 { server: 'Cowboy',
  date: 'Tue, 10 Mar 2015 06:26:27 GMT',
  connection: 'keep-alive',
  'strict-transport-security': 'max-age=631152000',
  'x-frame-options': 'DENY',
  'x-xss-protection': '1; mode=block',
  'x-content-type-options': 'nosniff',
  'content-security-policy': 'default-src \'self\'; connect-src \'self\'; font-src \'self\'; frame-src \'self\'; img-src https://* \'self\' data:; media-src \'none\'; object-src \'self\'; script-src \
'self\' \'unsafe-inline\' https://ssl.google-analytics.com https://www.google-analytics.com; style-src \'self\' \'unsafe-inline\';',
  location: 'http://gh-contractor-zcbenz.s3.amazonaws.com/atom-shell/dist/v0.21.0/node-v0.21.0.tar.gz',
  'content-type': 'text/html; charset=utf-8',
  'cache-control': 'no-cache',
  'set-cookie': [ 'request_method=GET; path=/; secure' ],
  'x-request-id': '81ca1402-fd67-4da2-bb76-7b12876e4768',
  'x-runtime': '0.013935',
  'transfer-encoding': 'chunked',
  via: '1.1 vegur' }
REQUEST redirect http://gh-contractor-zcbenz.s3.amazonaws.com/atom-shell/dist/v0.21.0/node-v0.21.0.tar.gz
REQUEST redirect to http://gh-contractor-zcbenz.s3.amazonaws.com/atom-shell/dist/v0.21.0/node-v0.21.0.tar.gz
REQUEST {}
REQUEST response end http://gh-contractor-zcbenz.s3.amazonaws.com/atom-shell/dist/v0.21.0/node-v0.21.0.tar.gz 302 { server: 'Cowboy',
  date: 'Tue, 10 Mar 2015 06:26:27 GMT',
  connection: 'keep-alive',
  'strict-transport-security': 'max-age=631152000',
  'x-frame-options': 'DENY',
  'x-xss-protection': '1; mode=block',
  'x-content-type-options': 'nosniff',
  'content-security-policy': 'default-src \'self\'; connect-src \'self\'; font-src \'self\'; frame-src \'self\'; img-src https://* \'self\' data:; media-src \'none\'; object-src \'self\'; script-src \
'self\' \'unsafe-inline\' https://ssl.google-analytics.com https://www.google-analytics.com; style-src \'self\' \'unsafe-inline\';',
  location: 'http://gh-contractor-zcbenz.s3.amazonaws.com/atom-shell/dist/v0.21.0/node-v0.21.0.tar.gz',
  'content-type': 'text/html; charset=utf-8',
  'cache-control': 'no-cache',
  'set-cookie': [ 'request_method=GET; path=/; secure' ],
  'x-request-id': '81ca1402-fd67-4da2-bb76-7b12876e4768',
  'x-runtime': '0.013935',
  'transfer-encoding': 'chunked',
  via: '1.1 vegur' }
REQUEST make request http://gh-contractor-zcbenz.s3.amazonaws.com/atom-shell/dist/v0.21.0/node-v0.21.0.tar.gz
gyp WARN install got an error, rolling back install
gyp
```

It seems like atom-shell is not fetchable over the proxy... Is there a way to work around this? Atom-Shells http response code is 302 (which is a location redirect). The location object (as seen in the last line) is the same as the original request uri...

As I am not allowed to do something on the proxy (Squid) I hope there is a workaround for this kind of problems.

Best regards,
Christian

**Comment by cyberis** [comment]
I am having the same problem with Atom 0.174.1 for Windows.  When I do some curls this is what I get:
```
$ curl -I https://atom.io/download/atom-shell/v0.21.0/node-v0.21.0.tar.gz
HTTP/1.0 200 Connection established

HTTP/1.1 302 Found
Server: Cowboy
Date: Tue, 10 Mar 2015 20:13:35 GMT
Connection: keep-alive
Strict-Transport-Security: max-age=631152000
X-Frame-Options: DENY
X-Xss-Protection: 1; mode=block
X-Content-Type-Options: nosniff
Content-Security-Policy: default-src 'self'; connect-src 'self'; font-src 'self'; frame-src 'self'; img-src https://* 's
elf' data:; media-src 'none'; object-src 'self'; script-src 'self' 'unsafe-inline' https://ssl.google-analytics.com http
s://www.google-analytics.com; style-src 'self' 'unsafe-inline';
Location: http://gh-contractor-zcbenz.s3.amazonaws.com/atom-shell/dist/v0.21.0/node-v0.21.0.tar.gz
Content-Type: text/html; charset=utf-8
Cache-Control: no-cache
Set-Cookie: request_method=HEAD; path=/; secure
X-Request-Id: 87f85c20-ee53-4521-84ca-9af13baa4375
X-Runtime: 0.012286
Via: 1.1 vegur

$ curl  https://atom.io/download/atom-shell/v0.21.0/node-v0.21.0.tar.gz
<html><body>You are being <a href="http://gh-contractor-zcbenz.s3.amazonaws.com/atom-shell/dist/v0.21.0/node-v0.21.0.tar
.gz">redirected</a>.</body></html>

```
If I do a wget on the redirected resource then I get the file. 
```
$ wget  http://gh-contractor-zcbenz.s3.amazonaws.com/atom-shell/dist/v0.21.0/node-v0.21.0.tar.gz
--2015-03-10 16:23:26--  http://gh-contractor-zcbenz.s3.amazonaws.com/atom-shell/dist/v0.21.0/node-v0.21.0.tar.gz
Resolving proxy.company.com (proxy.company.com)... 192.168.0.1
Connecting to proxy.company.com (proxy.company.com)|192.168.0.1|:3128... connected.
Proxy request sent, awaiting response... 200 OK
Length: 423657 (414K) [application/x-tar]
Saving to: `node-v0.21.0.tar.gz'

100%[==============================================================================>] 423,657     --.-K/s   in 0.1s

2015-03-10 16:23:27 (3.39 MB/s) - `node-v0.21.0.tar.gz' saved [423657/423657]
```
That's using Cygwin on the same computer that I am having trouble with. Is gyp unable to follow the redirect?

**Comment by brooklynb7** [comment]
The same issue with 0.187.0 on Windows 8.1

**Comment by michaelkrnac** [comment]
Same Problem here with 0.187 on Windows 7

**Comment by SoulBeaver** [comment]
Same issue with 0.188 on Windows 8.1

**Comment by wonneseufzer** [comment]
Same issue with 0.188 on Win7 behind a proxy. Would be great to have a workaround.

**Comment by Platzer** [comment]
Same Problem here with 0.189 on Windows 7

**Comment by curio** [comment]
Same here with 0.190 on Linux.

**Comment by tborychowski** [comment]
Same here with 0.190, Windows 8.1

**Comment by Platzer** [comment]
Same Problem here with 0.190 on Windows 7

**Comment by tborychowski** [comment]
seems to be related to https://github.com/atom/atom/issues/2963#issuecomment-90862226

**Comment by radum** [comment]
Same Problem here with 0.190 on Windows 7

**Comment by tborychowski** [comment]
seems related to https://github.com/atom/apm/issues/322

**Comment by gaving** [comment]
same issue with 0.196 on Windows 8.1 :gun:

**Comment by aureliojargas** [comment]
Thanks @gaving, https://github.com/atom/apm/issues/322#issuecomment-96430856 also fixed the issue here. (win7, Atom 1.4.0)

**Comment by damieng** [comment]
Looks like this was fixed by atom/apm#322 so closing here. Please re-open or re-file if you still experience issues.
